src/spirc.rs

```diff
@@ -513,38 +513,70 @@ impl SpircTask {
             self.state.set_position_measured_at(now);
         }
     }
+    fn consume_queue(&mut self) -> bool {
+        let current_index = self.state.get_playing_track_index() as usize;
+        if self.state.get_track()[current_index].get_queued() {
+            self.state.mut_track().remove(current_index);
+            return true
+        }
+        false
+    }
 
     fn handle_next(&mut self) {
         let current_index = self.state.get_playing_track_index();
+        let mut new_index = if self.consume_queue() {
+            current_index
+        } else {
+            current_index + 1
+        };
         let num_tracks = self.state.get_track().len() as u32;
-        let new_index = (current_index + 1) % num_tracks;
-
-        let mut was_last_track = (current_index + 1) >= num_tracks;
-        if self.state.get_repeat() {
-            was_last_track = false;
-        }
-
+        let continue_playing = if new_index >= num_tracks {
+            new_index = 0; // Loop around back to start
```

Comment:
Looking at this again, I'm going to move this out into it's own if statement and make it more obvious.